core/client/src/main/java/alluxio/client/file/FileInStream.java

```diff
@@ -404,4 +412,66 @@ private void updateBlockInStream(long blockId) throws IOException {
       mShouldCacheCurrentBlock = mAlluxioStorageType.isStore();
     }
   }
+
+  /**
+   * Seeks to a file position. Blocks are not cached unless they are fully read. This is only called
+   * by {@link FileInStream#seek}.
+   *
+   * @param pos The position to seek to. It is guaranteed to be valid (pos >= 0 && pos != mPos &&
+   *            pos <= mFileLength)
+   * @throws IOException if the seek fails due to an error accessing the stream at the position
+   */
+  private void seekInternal(long pos) throws IOException {
+    seekBlockInStream(pos);
+    checkAndAdvanceBlockInStream();
+    mCurrentBlockInStream.seek(mPos % mBlockSize);
+  }
+
+  /**
+   * Seeks to a file position. Blocks are cached if they are fully read. This is only called by
+   * {@link FileInStream#seek}.
+   * Invariant: if the current block is to be cached, [0, mPos) should have been cached already.
+   *
+   * @param pos The position to seek to. It is guaranteed to be valid (pos >= 0 && pos != mPos &&
+   *            pos <= mFileLength).
+   * @throws IOException if the seek fails due to an error accessing the stream at the position
+   */
+  private void seekInternalWithCachingIncompleteBlock(long pos) throws IOException {
+    boolean isInCurrentBlock = pos / mBlockSize == mPos / mBlockSize;
+    if (mShouldCacheCurrentBlock) {
+      // Cache the current block or read forward to pos.
+      readCurrentBlockTill(pos > mPos ? pos : mFileLength);
+    }
+    if (mPos == pos) {
+      return;
+    }
+
+    if (isInCurrentBlock) {
+      seekBlockInStream(pos);
+      mCurrentBlockInStream.seek(mPos % mBlockSize);
+    } else {
+      seekBlockInStream(pos / mBlockSize * mBlockSize);
+      checkAndAdvanceBlockInStream();
+      readCurrentBlockTill(pos);
+    }
+  }
+
+  /**
+   * Reads the current block till pos or the end of the current block if some conditions are met.
+   * No-op is pos <= mPos.
```

Comment:
NIT (try to be more explicit):
Reads till the file offset (mPos) equals pos or the end of the current block (whichever is achieved) if pos > mPos; otherwise a no-op.